cpp/include/cudf/table/row_operator3.cuh

```diff
@@ -0,0 +1,315 @@
+                        int const* depth                  = nullptr,
+                        order const* column_order         = nullptr,
+                        null_order const* null_precedence = nullptr)
+    : _lhs{lhs},
+      _rhs{rhs},
+      _nulls{has_nulls},
+      _depth{depth},
+      _column_order{column_order},
+      _null_precedence{null_precedence}
+  {
+  }
+
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the
+   * `rhs` table
+   */
+  __device__ bool operator()(size_type lhs_index, size_type rhs_index) const noexcept
+  {
+    int last_null_depth = std::numeric_limits<int>::max();
+    for (size_type i = 0; i < _lhs.num_columns(); ++i) {
+      int depth = _depth == nullptr ? std::numeric_limits<int>::max() : _depth[i];
+      if (depth > last_null_depth) { continue; }
+
+      bool ascending = (_column_order == nullptr) or (_column_order[i] == order::ASCENDING);
+
+      null_order null_precedence =
+        _null_precedence == nullptr ? null_order::BEFORE : _null_precedence[i];
+
+      auto comparator = element_relational_comparator{
+        _nulls, _lhs.column(i), _rhs.column(i), null_precedence, depth};
+
+      weak_ordering state;
+      thrust::tie(state, last_null_depth) =
+        cudf::type_dispatcher(_lhs.column(i).type(), comparator, lhs_index, rhs_index);
+
+      if (state == weak_ordering::EQUIVALENT) { continue; }
+
+      return state == (ascending ? weak_ordering::LESS : weak_ordering::GREATER);
+    }
+    return false;
+  }
+
+ private:
+  table_device_view _lhs;
+  table_device_view _rhs;
+  Nullate _nulls{};
+  null_order const* _null_precedence{};
+  order const* _column_order{};
+  int const* _depth;
+};  // class row_lexicographic_comparator
+
+struct preprocessed_table {
+  preprocessed_table(table_view const& table,
+                     host_span<order const> column_order,
+                     host_span<null_order const> null_precedence,
+                     rmm::cuda_stream_view stream);
+
+  operator table_device_view() { return **d_t; }
+  order const* column_order() { return d_column_order.data(); }
+  null_order const* null_precedence() { return d_null_precedence.data(); }
+  int const* depths() { return d_depths.data(); }
+  [[nodiscard]] bool has_nulls() const { return _has_nulls; }
+
+ private:
+  using table_device_view_owner =
+    std::invoke_result_t<decltype(table_device_view::create), table_view, rmm::cuda_stream_view>;
+
+  std::unique_ptr<table_device_view_owner> d_t;
+  rmm::device_uvector<order> d_column_order;
+  rmm::device_uvector<null_order> d_null_precedence;
+  rmm::device_uvector<size_type> d_depths;
```

Comment:
What's the average overhead for the allocations/copies for the preprocessing?

One optimization here would be to fit all of the `table_device_view` data and all of the other preprocessed device data into a single device allocation with a single H2D copy.

Reply:
The overall time it takes to copy and sync is in 10s of microseconds.

It should be relatively easy to combine d_column_order, d_null_precedence, and d_depths into a single allocation. I don't think a table_device_view constructor exists that lets us put data into pre-allocated storage.